include/ginkgo/core/base/batch_multi_vector.hpp

```diff
@@ -0,0 +1,813 @@
+          batch_size_{compute_batch_size(matrices)},
+          values_(exec, compute_num_elems(batch_size_))
+    {
+        for (size_type i = 0; i < this->get_num_batch_entries(); ++i) {
+            auto local_exec = matrices[i]->get_executor();
+            exec->copy_from(
+                local_exec.get(), matrices[i]->get_num_stored_elements(),
+                matrices[i]->get_const_values(),
+                this->get_values() + this->get_size().get_cumulative_offset(i));
+        }
+    }
+
+    /**
+     * Creates a BatchMultiVector matrix by duplicating BatchMultiVector matrix
+     *
+     * @param exec  Executor associated to the vector
+     * @param num_duplications  The number of times to duplicate
+     * @param input  the vector to be duplicated.
+     */
+    BatchMultiVector(std::shared_ptr<const Executor> exec,
+                     size_type num_duplications,
+                     const BatchMultiVector<value_type>* input)
+        : BatchMultiVector(
+              exec, gko::batch_dim<2>(
+                        input->get_num_batch_entries() * num_duplications,
+                        input->get_common_size()))
+    {
+        size_type offset = 0;
+        for (size_type i = 0; i < num_duplications; ++i) {
+            exec->copy_from(
+                input->get_executor().get(), input->get_num_stored_elements(),
+                input->get_const_values(), this->get_values() + offset);
+            offset += input->get_num_stored_elements();
+        }
+    }
+
+    /**
+     * Creates a BatchMultiVector matrix by duplicating Dense matrix
+     *
+     * @param exec  Executor associated to the vector
+     * @param num_duplications  The number of times to duplicate
+     * @param input  the vector to be duplicated.
+     */
+    BatchMultiVector(std::shared_ptr<const Executor> exec,
+                     size_type num_duplications,
+                     const matrix::Dense<value_type>* input)
+        : BatchMultiVector(
+              exec, gko::batch_dim<2>(num_duplications, input->get_size()))
+    {
+        size_type offset = 0;
+        for (size_type i = 0; i < num_duplications; ++i) {
+            exec->copy_from(
+                input->get_executor().get(), input->get_num_stored_elements(),
+                input->get_const_values(), this->get_values() + offset);
+            offset += input->get_num_stored_elements();
+        }
+    }
+
+    /**
+     * Creates a BatchMultiVector matrix with the same configuration as the
+     * callers matrix.
+     *
+     * @returns a BatchMultiVector matrix with the same configuration as the
+     * caller.
+     */
+    std::unique_ptr<BatchMultiVector> create_with_same_config() const
+    {
+        return BatchMultiVector::create(this->get_executor(), this->get_size());
+    }
+
+    /**
+     * @copydoc scale(const BatchMultiVector *)
+     */
+    void scale_impl(const BatchMultiVector<ValueType>* alpha);
```

Comment:
Without the virtual stuff, we don't need the two staged implementation `scale->scale_impl` anymore. (Same for the other ones)